synapse/util/caches/stream_change_cache.py

```diff
@@ -271,6 +277,13 @@ def entity_has_changed(self, entity: EntityType, stream_pos: int) -> None:
         if stream_pos <= self._earliest_known_stream_pos:
             return
 
+        # Any change being flagged must be ahead of any current token, otherwise
+        # we have a race condition between token position and stream change cache.
+        # NOTE: this checks for equal to allow for a process persisting an event to
```

Comment:
Not super pleased with this perhaps a better way would be to add a flag to explicitly enable exact matches for writer instances?